### ISSUE TITLE
Speedez - Forward  transitioned child account balance to parent account

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#1323]
In GitLab by @kdjstudios on Jan 14, 2019, 09:02

**Submitted by:** Jo Ann Browne <joann@speedez.com>

**Helpdesk:** http://www.servicedesk.answernet.com/profiles/ticket/6578018

**Server:** External

**Client/Site:** Speed Ez

**Account:** 1002

**Issue:** 

Problems have occurred with the Parent Account 1002 for Conviva Care Solutions. It appears that if an account had a past due balance, it did not carry it forward to be billed and there is no way to print out any statement before the Parent account was established. 

Am I missing something?  Is there anyway I can get copies of the past due bills?  It adds up to $1387.26. 

On another note, I had ask you about the Sales Tax NOT billing on the child account (only a Sub-total shows). U were going to check into WHY. In this case with 1002 it did not charge them sales tax at all (another $275.20 lost). 

This is quite serious and I am really worried. I will send you some pictures on a separate email.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 14, 2019, 09:02

Jo Ann wrote:

> Here are the photos showing the Balances Due and the other 3 photos show how the accounts were billed but no where do you see the Past-Due amounts. How do I print out bills for those 3 accounts. 

![IMG_8110](/uploads/acc43a6d0a9df941c49285f722791554/IMG_8110.jpg)

![IMG_8111](/uploads/a66b6a9b17c0de8cd1063de46f2ed8b9/IMG_8111.jpg)

![IMG_8112](/uploads/6ad726c6f3eab1526ae9eecaded11476/IMG_8112.jpg)

![IMG_8109](/uploads/704f4b3cb76c85a644e23ccaddfc4cd4/IMG_8109.jpg)


Jo Ann Wrote:

> Another question!  If this company pays the past due charges - where do we enter the payment at?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 14, 2019, 09:05

@smasih @amishra @pchaudhary @tim.traylor 

Please look into the questions and concerns from Jo Ann above and provide explanations for each thank you.

EDIT: From what I can tell is when the accounts were made as child accounts to the parent account:
- The child account previous balance due did not carry over to the parent accounts invoicing or balance.
- Since (#1155 is still an issue) on any child account there is no way to pay off it's balance?
- The parent and child accounts all appear to be setup to charge taxes, but I do not see any taxes on the parent account's last invoice?
- #1296 is still needing to be reviewed by Gary to determine if this is a bug or an enchantment and if we need to charge the client.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 14, 2019, 09:23

I wrote:

> Thank you for notifying us. We have informed the support team and will have them research this and follow up accordingly. Thank you for the screenshots of the accounts that show the child accounts have balances that did not get billed up into the parent account. Also in regards to the missing taxes; I took a quick look and everything appears to be setup correctly. We will need to have the support team look further into that concern and follow up accordingly. In regards to the separation of taxes by sub account, I will send an update on that thread later today as it has been in our recent discussions with support too. Lastly, with making payments on the child account balances, I will check with the support team if those balances will be moved to the parent account or how we will need to make payments on those balances accordingly.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 15, 2019, 01:18

@kdjstudios 


> If an account had a past due balance, it did not carry it forward to be billed and there is no way to print out any statement before the Parent account was established.

* Yes, if an account had a past due balance, it'll not carry it forward to be billed on the parent account. It is also true that currently, there is no way to print out any statement before the Parent account was established.

> Is there any way I can get copies of the past due bills?

* We can make it possible to print out the previous invoices of the account on its own show page as same as parent account by updating the code accordingly.  

 

> I had to ask you about the Sales Tax NOT billing on the child account (only a Sub-total shows). U were going to check into WHY. In this case, with 1002 it did not charge them sales tax at all (another $275.20 lost)

* Currently, SAB fails to add the taxes basis on the child account. #1296 


> Since (#1155 is still an issue) on any child account there is no way to pay off it's balance?

* No, you can pay the remaining balance from the account page.  
  * Can make a credit card payment.
  * Can make a credit invoice.

* Only, the manual payment option is not there for use.

> The parent and child accounts all appear to be setup to charge taxes, but I do not see any taxes on the parent account's last invoice?

* We dug into this & after a long struggle we found that the account has no selected tax at the point of invoice creation which leads to no taxes added on the invoice.  


Hope this helps.
Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 15, 2019, 12:03

Marking this for Review with Gary and @tim.traylor . We will need further clarification on how these should be handled. 

- The parent and child accounts all appear to be setup to charge taxes, but I do not see any taxes on the parent account's last invoice?

 - We dug into this & after a long struggle we found that the account has no selected tax at the point of invoice creation which leads to no taxes added on the invoice.

> - This makes no sense, I see on every single child account it was marked to charge taxes on all billable activities, fees, etc. I also see that on each child account it has assigned to it a tax also. The parent account also has a tax associated with it. Please clarify what you mean.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 15, 2019, 23:41

@kdjstudios 

> This makes no sense, I see on every single child account it was marked to charge taxes on all billable activities, fees, etc. I also see that on each child account it has assigned to it a tax also. The parent account also has a tax associated with it. Please clarify what you mean.

* I think you are confused with all of the above stuff. Let me clear: 
  * There are multiple child accounts of account **Conviva Care Solutions Parent Account** & all of them have individually selected taxes on each but as per the current bug on SAB (noticed on #1296 ) none of them are applied individually basis on the per account subtotal.  
  * If the parent account **Conviva Care Solutions Parent Account** contains the tax at the time of invoice creation then the total amount (*total of the subtotal of all the accounts*) is calculated for the tax on the invoice. But as I mentioned earlier that at the time of invoice creation the parent account's tax *Texas state tax* is not selected so none of the tax was applied to the last invoice of the account.

Hope it clears.

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 16, 2019, 10:09

@pchaudhary 

Thank you, Can you please confirm the following needs to occur:

1. Find which user and when added the Texas state tax to the parent account.
2. Confirm that since they now have the Texas state tax on the parent account, all child account taxes need will be sum into the parent account. (Until the #1296 fix is released)


My thoughts:
- Each account should only charge applicable taxes to its account. Such if a child account is setup for taxes and the parent account is not. there should still be the child account taxes on the parent account's invoice, but only on the child account fees. Please insure this scenario is tested in #1296 . I will also post there too.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 16, 2019, 10:16

I wrote:

> The support team is still looking into this. We do have a partial update as to what they are finding so far. The support team has informed me that SA Billing currently has to have the taxes enabled at the parent account and child account in order for the items to be billed with Taxes onto a parent account. The support team found at the time the parent accounts invoice was generated that taxes were not enabled on the parent account, but they were enabled on child accounts. The parent account then override the taxes as disabled when generating the parent account's invoice. 
> 
> We are discussing this further internally on what will need to be adjusted; as we do understand that sometimes a parent account may not need to bill for taxes on it's charges, but child accounts will want to bill for taxes as well. This actually goes along with your previous request about having independent tax charges on child accounts; when billing into a parent account. 
> 
> We will keep you updated as progress is made. Please let us know if you have any questions or concerns.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 17, 2019, 03:07

@kdjstudios 


> Find which user and when added the Texas state tax to the parent account.

* There is no such mechanism to find the user who updated that account.  
* The thing on which we can confirm are:  
  * The taxes was added on 13^th Jan
  * Because we checked with 12^th Jan dump & there was no taxes on the Site.


> Each account should only charge applicable taxes to its account. Such if a child account is setup for taxes and the parent account is not. there should still be the child account taxes on the parent account's invoice, but only on the child account fees.

* Agrred.
* Do we need to manually added the tax on each of the child account's subtotal on which the tax is selected?

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 17, 2019, 10:04

After discussion with Gary:

We need to prompt to keep balance or forward to the parent account. When a user selects a parent account from the drop down. 

Child account Taxes (any charges, holiday late, etc) (if enabled) need to go to the parent regardless if the parent account has taxes enabled.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jan 22, 2019, 06:14

assigned to @pchaudhary

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 24, 2019, 02:38

@kdjstudios 

Thanks for the update. 
Would you please help us to provide the full process flow with the mockup so that we can have a better understanding of the requirement & then we can estimate it accordingly.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jan 24, 2019, 04:02

old bug, and we will be writing addition code to fix it marking this as enhancement as well.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 24, 2019, 10:48

marking for review with @tim.traylor to setup the mock ups and process flow.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Jan 25, 2019, 06:40

@kdjstudios Hey Kyle - can you please bring this up on our Tuesday AM meeting?  I'd like to have all of us discuss it.

Also, Sumeet is going to check if we currently are able to process a payment against a child account, which I guess would make Gary's request possible:![image](/uploads/e20489519289db68899ea9cc4a708dcd/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 25, 2019, 09:37

@tim.traylor 

I believe payments on child account was already researched above http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1326#note_32800. It was stated that we can only make electric payments via Echeck or CC. The post payment functionality is not available, as it is only accessible from the Payment history screen which is hidden on child accounts.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 29, 2019, 06:14

@kdjstudios @tim.traylor 

While testing on one of the other issues, we found that there is one another scenario under the parent-child relationship which is failing at the moment.  

* If there is a child account having Quarterly Billing Cycle & the parent account do have a Monthly billing cycle then while creating an invoice of the Parent account, the child's fees will also be calculated on that.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jan 29, 2019, 07:31

> Also, Sumeet is going to check if we currently are able to process a payment against a child account, which I guess would make Gary's request possible:

We can make a CC and echeck payment on the child account. Although the account selection dropdown is empty but still if I proceed payment is posted on child account.      
![image](/uploads/b6c964087a617895512ff82033abc2b0/image.png)

We cannot make any kind of payment on child account through **External/Customer Portal**

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 29, 2019, 10:41

@pchaudhary @smasih @amishra @tim.traylor 

If there is a child account having Quarterly Billing Cycle & the parent account do have a Monthly billing cycle then while creating an invoice of the Parent account, the child's fees will also be calculated on that.

> Should we even be allowing for different billing cycles (Or Billing Periods?) on parent child accounts? This seems like something we should not be allowing.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 29, 2019, 10:44

We can make a CC and echeck payment on the child account. Although the account selection dropdown is empty but still if I proceed payment is posted on child account.  

> @tim.traylor This seems like a bug that we allow for a payment without specifying the account.

We cannot make any kind of payment on child account through **External/Customer Portal**

> @tim.traylor I think this will need to be reviewed with Gary to determine how he wants the program to function.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Jan 30, 2019, 09:32

For Gary Review:
If we allow accounts to be made into a child account while it still has a balance then we either need to move the balance from the new child account to the parent account (as some sort of staged fee) OR we don't move the balance and we allow payments to be posted against child accounts.

ALternately, we could simply not allow making an account a child account if it has a non-zero balance.  In this case they would need to apply payments or adjustments and get it to zero before allowing them to change the account type.  

Also, @smasih Sumeet raised the question of handling the security deposits. The team is going to check and see if we can leave the SD on the child account and have everything work properly, or if we need to transfer it to the parent account.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 31, 2019, 09:43

The balance of the child account will need to be shown/forwarded onto the parent account. Please provide an estimate for this.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 4, 2019, 11:29

@amishra @pchaudhary @smasih 

May we please have an estimate?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 11, 2019, 09:10

@smasih @amishra @pchaudhary @tim.traylor 

Its been over a week and we still do not have an estimate on this ticket. This is marked as Important, may we please have an estimate.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 25, 2019, 12:30

changed title from **Speedez -** to **Speedez -{+ Forward  transitioned child account balance to parent account+}**

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 14, 2019, 11:07

@smasih @amishra @pchaudhary @tim.traylor @rchauhan 

As it was pointed out in #1399 that the team has not been manually moving the child account balance to parent accounts when they are manually moving the accounts; this has now resulted in child accounts that still have a balance. So we will need in addition to the estimate to fix the underlining issue; to also have a list of all child accounts that currently have non zero balances.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on May 31, 2019, 04:11

@tim.traylor @kdjstudios 

> So we will need in addition to the estimate to fix the underlining issue. 

* We need to discuss the flow of the process which we use to make an account as the child account.  


> to also have a list of all child accounts that currently have non zero balances. 

* Please also find the list of the child accounts of the production servers which are having the balance - 

  * Internal - [child_accounts_with_balance_internal.csv](/uploads/21b0082a6f87df1f39fbb644edaa9824/child_accounts_with_balance_internal.csv)
  * External - No Account



Thanks

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Jun 27, 2019, 10:25

@smasih @pchaudhary 
The need is to take a standard (top-level, non-child) account and turn it into a child account of an existing account.

When this happens we need to move the balance from the child account up to the new parent account so it is not lost.  We will do this by applying a credit payment to the child account and a debit staged fee to the parent account.

As we discussed, there is no reasonable way to do this balance transfer to the parent and still retain the aging.  So we need to notify the user that the aging is going to be lost when they do this.

Question:
*  If we're doing a credit payment on the child account, how do we get that finalized w/out a billing cycle?  Does it just get finalized during the next billing cycle closure?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Aug 6, 2019, 10:52

@tim.traylor @smasih 

I believe, we have had about 4-5 tickets to my recollection within just the last year or so to manually move accounts and their balances into child accounts on different customers. Based upon the estimate in #1542 we are spending 2 hours roughly on each of those tickets; that's about 10 hours already spent. We need an estimate on this ticket as I would expect fixing the actual issue will save us time in the long run.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Aug 7, 2019, 05:04

@kdjstudios [Tim](@tim.traylor) 

Thanks for the update. 
There are multiple items which need to take care at the time of Account transfer.

1- Child account balance should be added properly as the Parent account staged fee.  
2- The Child account balance has to be zero before the transfer. There should be a transaction to make the account balance zero. 
  * a) Child account balance is positive:
      * Add a manual payment (*on child account*) for the amount of the same as the child's account balance.
  * b) Child account balance is negative:
      * Need to create an Invoice of the positive amount of the account's balance so that after the finalization of the invoice, the account's balance will be zero.  


>**Tim's comment**:  As we discussed, there is no reasonable way to do this balance transfer to the parent and still retain the aging.  So we need to notify the user that the aging is going to be lost when they do this.

* I am not sure that why did you mention this thing because as per my best understanding the aging will not reflect any miscalculation if we follow the above steps. Please correct me if I am not getting your words somewhere.

**Note** - We also would like to highlight one concern which should be resolved before we further move into the development on this ticket. The issue was also highlighted by [Tim](@tim.traylor) on his http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1326#note_43140 that if the child account has the payment which is not printed on any of the invoices yet, then there is still the Delete button to delete that payment.  
> We can resolve this issue by only adding the additional conditional operator (if) on the delete button shows on the payment index page & also on the controller level validation.


Hope this helps.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Aug 7, 2019, 05:27

@pchaudhary Hi Piyush,  if the balance on the child account is not current and we move the balance to the parent, we will have 'paid off' the child invoice to zero and thus cleared the aging.  When the parent account's next invoice is generated the previous child account balance becomes a current due on the parent.  Correct?

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Aug 7, 2019, 05:51

@kdjstudios Kyle,  I know we discussed the mechanics of moving the balances with Gary, but I don't see it on this ticket.  Can you pls identify which ticket has that discussion so we can verify the steps?  Thx

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Aug 7, 2019, 14:49

@tim.traylor 

I think it was earlier on this ticket http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1326#note_32994

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Aug 8, 2019, 24:16

added 2h 30m of time spent

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Aug 8, 2019, 01:01

@tim.traylor 

> Hi Piyush,  if the balance on the child account is not current and we move the balance to the parent, we will have 'paid off' the child invoice to zero and thus cleared the aging.  When the parent account's next invoice is generated the previous child account balance becomes a current due on the parent.  Correct?

* Yes, correct. In this case, the child's account whole aging (0,30,60,90,120+) will be moved to the current aging of the parent after the parent account's new invoice finalization.

@kdjstudios 

Thanks for the pointing towards the discussion with Gary.

It would be great if you can elaborate it more by providing some more details.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Aug 9, 2019, 01:27

added 30m of time spent

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Aug 22, 2019, 12:16

@pchaudhary 

I am confused on what more details you are needing? Everything is addressed in previous posts.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Sep 18, 2019, 07:25

Will discuss with Sumeet and post an estimate tomorrow.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 26, 2019, 09:54

@smasih @amishra 

Please outline the specs from all previous posts into a consolidated post. We can then review this to ensure your understanding.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 27, 2019, 12:31

Gary reviewed this today and will want @jredfield to help with creating the mock ups, design, and estimate.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Nov 14, 2019, 23:09

@kdjstudios @jredfield 

Thank you for the update.

Any update on this, as we are still waiting for the mockups and design. This ticket is also labeled as important.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 15, 2019, 10:52

@ranand @amishra @smasih 

I bleieve this is already detailed out in this post http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1326#note_43140 and this post http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1326#note_45065 . If there is still something missing please let us know, but to my knowledge this is all back end changes that require no mockup.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Nov 22, 2019, 04:35

unassigned @pchaudhary

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Nov 29, 2019, 06:13

@kdjstudios @jredfield

As per my understanding of this ticket, we need to make provisions when we are making an existing non-child account to the child account. For that, we will take the following steps:

**CASE 1: The child account has a positive account balance**

To balance out the child account we have two options i.e

* Option 1: Add a positive staged fee on the child account and create an invoice.
* Option 2: Add a manual payment on the child account and create an invoice

For the parent account, we will add a staged fee with the child account balance which will come on the next invoice of the parent account.

**CASE 2: The child account has a negative account balance**

To balance out the child account we will add a negative staged fee on the child account and create an invoice.

For the parent account, we will add a staged fee with the child account balance which will come on the next invoice of the parent account.


**If my understanding of the ticket is correct then before providing an estimate to this ticket, I have some queries**

* Which option i.e Option 1 or Option 2 should we go with, in case of positive balance on the child account?
* What we will do with the Security Deposit of the child account?
* What we should do with the taxes, late charges for the child account?
* Which staged activity we will use to balance the accounts?

**NOTE:** Taxes functionality for the parent-child account is fixed in the #1296 issue.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 29, 2019, 12:11

@ranand @smasih @amishra 

Both cases should use the same steps, So that means we need to use option 1 on the case 1.

The SD should remain on the child accounts as the SD should already be processed on child accounts correctly according to #28 

Taxes, holidays, and late charges should remain the same as #1296 and #1327 should be addressing the issues with parent/child issues. 

@jredfield 

For determining the staged fee to use, we may need to check with Gary first, I bet we will have to check with each site/client, or setup a new 'static' fee to use for all.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Dec 2, 2019, 23:21

@kdjstudios @jredfield 

We need clarification for one more scenario i.e

* while creating an invoice on the child account to make the account balance zero(0) and if the Billing cycle is finalized. In this case, what do we should do?

We can think of two options to solve it

**Option 1:** We can throw a warning and say something like.."Please create a new billing cycle before proceeding because the account has a balance."  

**Option 2:** We can create the invoice on the finalized billing cycle(current billing cycle).


**And if you have any other possible solution for this then, please suggest.**


## Developer Estimate
|Size|Lower Time Bound|Variability|Upper Time Bound|Testing|Time for Testing| Time for Dev & Testing|
|-|-|-|-|-|-|-|
|Medium-1|6|Worked with Concepts, implementation ok|8|High Impact|4|12|

**Estimate can be different if we go with other(except option 1 & 2) possible solution**  

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 3, 2019, 12:30

@amishra @jredfield 

I do not believe we should be adding invoices to finalized billing cycles, as that could effect historical reporting. I believe we will want to do option 1.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 4, 2019, 13:07

Marking this for approval, as we need to have this bug/loophole fixed.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Dec 5, 2019, 24:26

@kdjstudios @jredfield 

Thank you for approval and conformation.

* Please provide the verbiage for option 1.
> For determining the staged fee to use, we may need to check with Gary first, I bet we will have to check with each site/client, or set up a new 'static' fee to use for all.
* Also, I am still waiting for your response. What **stage fee** we need to use for credit/debit account balance while transfer.
 
We have started the development and while development we found another scenario.

* If an account has a draft invoice. In this case, what we need to do
  * Should we warn the user that the account has already a draft invoice, and ask him to perform either finalize or delete/revert the invoice? Because if we make the account as a child then there is no way to find the draft invoice as we do not show the draft invoice button on the child account.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 5, 2019, 09:35

* Please provide the verbiage for option 1.

> "This account currently has a balance and in order to move it to a child account, the balance will need to be forwarded to the parent account. Please create a new billing cycle to allow for SA Billing to automatically populate the needed transfers as staged fees." - We will want to review this with Gary for his input, before it goes live. 

* Also, I am still waiting for your response. What **stage fee** we need to use for credit/debit account balance while transfer.

> Adding to review with Gary on this. I wonder if we should allow for the user to pick the staged fee from a drop down list during this process, so that way any client/site can choose accordingly. 

* If an account has a draft invoice. In this case, what we need to do. Should we warn the user that the account has already a draft invoice, and ask him to perform either finalize or delete/revert the invoice? Because if we make the account as a child then there is no way to find the draft invoice as we do not show the draft invoice button on the child account.

> Child accounts should never have draft invoices to start with. If we are allowing users to move an account into the child status while it has a draft invoice, this is where we need to present the error. Please confirm if we are allowing this, if so we will need to open a separate ticket to address that.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 9, 2019, 16:20

For Gary:

We will want to review the verbiage for the warning/error notice.

Also we will want to review how to transfer the account balance from child to parent. We already know it will be a staged fee.
- Should the users select the billing code from the drop down list to select the staged fee.
- Should SAB have a static code for all sites to use. 
- Any other solutions?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 17, 2019, 10:09

Verbaige for warning notice: "This account is now (DATE) and the balance is now moved to the parent account. The aging will be lost/moved also."

Transfer balance code that will be used to zero out the child account and then add the balance to the new parent account.

This will be a static (system) code that will be used by all sites. Please create this code and make it non-editable on the site page. 

Can we move the dates to keep the aging on this balance? If possible please provide a separate estimate for this.

The user permissions to do this will be 05 or better. on external will be manager and above.

EDIT: please provide an updated estimate it the previous one is no longer valid.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Dec 18, 2019, 03:49

assigned to @amishra

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Dec 18, 2019, 04:15

@kdjstudios @jredfield 

Thank you for the update.

> Verbiage for warning notice: "This account is now (DATE) and the balance is now moved to the parent account. The aging will be lost/moved also."

* This message/warning we need to show when we select the parent account dropdown from the account edit page?

![image](/uploads/3845955dd0255285250ebba9589c3fa5/image.png)

* Is the same message/warning we need to show when the user needs to create a new billing cycle to make as a child account? if the billing cycle has a finalized stage.(See [comment](http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1326#note_52341))

> Transfer balance code that will be used to zero out the child account and then add the balance to the new parent account.
> This will be a static (system) code that will be used by all sites. Please create this code and make it non-editable on the site page.

* What is the fee name, code?

> Can we move the dates to keep the aging on this balance? If possible please provide a separate estimate for this.
* I did not understand. It would be great if you can elaborate.

> The user permissions to do this will be 05 or better. on external will be the manager and above.

* Understood.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 18, 2019, 09:52

This message/warning we need to show when we select the parent account dropdown from the account edit page?
> I believe we will only want to display that message if there was a balance and not once the click the drop down, but once they save the account. As that is when we will be doing the processing to move the balance correct?

Is the same message/warning we need to show when the user needs to create a new billing cycle to make as a child account? if the billing cycle has a finalized stage.(See [comment](http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1326#note_52341))
> Verbaige: "We are unable to process this, as you will need to create a new billing cycle. This is because the account already has a balance that will need to be transferred to the parent account. Once you have created the new billing cycle please update this account."

What is the fee name, code?
> Lets use the following, until we can confirm with Gary. Name: Parent - Child Account Balance Transfer; Code: "7050"

I did not understand. It would be great if you can elaborate.
> Example, we have an account with a past due balance of 90 days. When we transfer that to the parent account, can we also transfer that it is 'late' or part of the aging, so that way the new parent account will show on the aging and AR accordingly?

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Dec 19, 2019, 05:41

@kdjstudios @jredfield 

Thank you for the update.

> I believe we will only want to display that message if there was a balance and not once the click the drop down, but once they save the account. As that is when we will be doing the processing to move the balance correct?

As per my understanding, we will only show the warning if the user selects the account with the balance from the dropdown and actual processing will happen on the submit of the form. Is it correct?

> Example, we have an account with a past due balance of 90 days. When we transfer that to the parent account, can we also transfer that it is 'late' or part of the aging, so that way the new parent account will show on the aging and AR accordingly?

The above comment means that you want us to keep the aging of the transferred balance. Currently, we can't figure out a way to do it. Because once we create the invoice for the adjustment on the child account we will be clearing the invoice balances and their agings as well. 

While developing this we got another scenario; i.e if the child account has recurring fees and its current billing cycle is finalized at the time of transfer then the new billing cycle will be created and after that, the adjustment invoice will be created. Now, the adjustment invoice will have all the recurring fees of the child account.

To solve this, we need an additional 3 hours of development.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 19, 2019, 11:20

As per my understanding, we will only show the warning if the user selects the account with the balance from the dropdown and actual processing will happen on the submit of the form. Is it correct?
> Yes.

The above comment means that you want us to keep the aging of the transferred balance. Currently, we can't figure out a way to do it. Because once we create the invoice for the adjustment on the child account we will be clearing the invoice balances and their agings as well.
> Thank you, if we are unable to transfer the aging, we will need to display that as part of the warning. so the updated Verbiage for warning notice: "This account is now (DATE) and the balance is now moved to the parent account. The aging on this balance will be lost also."

While developing this we got another scenario; i.e if the child account has recurring fees and its current billing cycle is finalized at the time of transfer then the new billing cycle will be created and after that, the adjustment invoice will be created. Now, the adjustment invoice will have all the recurring fees of the child account.
> Was that not already addressed in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1326#note_52341 where we said to use option 1?

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Dec 20, 2019, 02:34

@kdjstudios @jredfield 

> While developing this we got another scenario; i.e if the child account has recurring fees and its current billing cycle is finalized at the time of transfer then the new billing cycle will be created and after that, the adjustment invoice will be created. Now, the adjustment invoice will have all the recurring fees of the child account.
>
> > Was that not already addressed in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1326#note_52341 where we said to use option 1?

Yes, we have addressed it in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1326#note_52341, but what we did not address was the recurring fees on the account. When we will create the new invoice under the new billing cycle all the recurring fees on the child account will be added to the new invoice, which we don't want. 
So, for the adjustment invoice, we will allow having only one line item i.e the staged fee. For the adjustment invoice, we need to restrict any other line item to added in it. 

For this, it will take 3 more effort into development.

## Developer Estimate
|Size|Lower Time Bound|Variability|Upper Time Bound|Testing|Time for Testing| Time for Dev & Testing|
|-|-|-|-|-|-|-|
|Medium-1|8|Worked with Concepts, implementation ok|11|High Impact|4|15|

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 20, 2019, 09:58

@jredfield 

We may need to discuss this, As there may also be other fees that need to be added, late fees, staged, fees, SD fees, etc?

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Jan 8, 2020, 06:29

Holding off on approval until we discuss on Thursday.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 9, 2020, 08:57

We discussed how each type of fee/charge should be handled. Below are our thoughts. 

1. Staged fee for the account balance transfer need to show up. Taxes should not be applied to this staged fee (Since it has already been charged).
2. Security deposit fees need to keep on the child accounts.
3. Staged fees can stay on the child account.
4. Late fees will not be assessed while we do the transfer balance on the child and parent account. But late fees will be assessed regular once a standard billing cycle invoice is generated. 
5. Holiday Fees will not be charged, As these will be charged on the standard billing cycle invoice. 
6. Base rate/recurring fees, This will also not be charged, as these will be charged on the standard billing cycle invoice. 
7. Taxes will not be charged, As these will be charged on the standard billing cycle invoice.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 9, 2020, 08:58

Moving back to estimate, in case that needs to be adjusted.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jan 9, 2020, 22:26

@kdjstudios @jredfield 

Thank you for the update.

For developing this functionality the estimate is the same as given above `15 hrs`.


Please approve this so that we can start the development.

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Jan 10, 2020, 06:06

Approved. Please give this high priority.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jan 14, 2020, 06:08

@kdjstudios @jredfield 

Thank you for your approval.

We have almost completed the development. I will let you once ready for your testing.

We have some concern:

* In my above [comment](http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1326#note_52477), the last point if an invoice already has a draft stage then what do we need to do in this case? 

* For now, We have added Additional Fee with name "Parent-Child Account Balance Transfer" and code "7050". Do we need to change this?

* This additional should not be updated by anyone because if we will give them permission to update this then we are unable to find them which fee we need to add at the time of balance transfer. (**, For now, nobody can't change the Fee name from the site page and activate/deactivate as well.**)
![image](/uploads/4a8a30199d50d7b56e699d4278c340e9/image.png)

* Do we need to send an email when we transfer the balance/make as a child account to the user or anyone?
* Do we need to track the details of how much balance transfer into the parent account & when it was transferred etc..?

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 14, 2020, 10:50

* For now, We have added Additional Fee with name "Parent-Child Account Balance Transfer" and code "7050". Do we need to change this?
> This is good for now.

* This additional should not be updated by anyone because if we will give them permission to update this then we are unable to find them which fee we need to add at the time of balance transfer. (**, For now, nobody can't change the Fee name from the site page and activate/deactivate as well.**) ![image]\
> no one but admins should be able to adjust this fee. 

* Do we need to send an email when we transfer the balance/make as a child account to the user or anyone?
> Just the notice in the UI for now. We may want to add the email notice later. 

* Do we need to track the details of how much balance transfer into the parent account & when it was transferred etc..?
> Yes.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Jan 14, 2020, 23:59

@kdjstudios @jredfield 

> Just the notice in the UI for now. We may want to add the email notice later.

We need the verbiage for the notification.

> Do we need to track the details of how much balance transfer into the parent account & when it was transferred etc..?

> Yes.

It will take an additional 1.5 hours for this.

## Developer Estimate
|Size|Lower Time Bound|Variability|Upper Time Bound|Testing|Time for Testing| Time for Dev & Testing|
|-|-|-|-|-|-|-|
|X-Small|1|NA|1|Low Impact|0.5|1.5|


Thanks

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Jan 15, 2020, 05:52

Yes please complete this.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jan 15, 2020, 06:03

@kdjstudios @jredfield 

Thank you for your approval.

@kdjstudios Please provide the verbiage for the notification.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Jan 20, 2020, 21:16

@kdjstudios @jredfield 


> Do we need to track the details of how much balance transfer into the parent account & when it was transferred etc..?

> Yes.

Please provide the verbiage for the notification.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 21, 2020, 08:48

@ranand @jredfield 

I think the below would be a sufficient notification. We will I am sure want to confirm this with Gary. 

"By confirming this move. The [account name] will become a child account to [account name]. This will result in the existing balance of [balance amount] to be transferred to the new parent account via staged fees. Please do not remove or alter these staged fees."

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Feb 3, 2020, 04:14

@kdjstudios @jredfield 

> no one but admins should be able to adjust this fee.

* I don't think we need to update the **name** by the Admin because when Admin will change the name we will unable to find that Staged Fee at the time of invoice creation when we move the balance.

**Or**

* We can make disable field of the **code** so that admin can change the name, but not code(**For now, Admin can only change the fee code, not fee name.**).

Although, We have completed the development & setup branch for your testing on the dev server Please test it.
Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 5, 2020, 14:44

@amishra 

The fee name or code should not be able to be changed by anyone except for admins. 

I just attempted to test this and think we may have switched branches because this does not appear to be working for me.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Feb 7, 2020, 24:29

@kdjstudios @jredfield 

Thank you for the update.

If we will make change a fee editable for admin.  when admin has changed the name/code and we will transfer the amount on the parent account then unable to find the fee which may cause the issue.

So we have two options to do:

1. No one can change the fee.
2. if the fee is not found/present on the site level then we need to add the fee name with
code `7050` on the site level inside the additional fee section. and notify on the UI like "Please add the fee name with code `7050` on the site level inside the additional fee section before transfer the account balance into the parent account."

Please confirm which options we should go-ahead.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 7, 2020, 09:56

@amishra @ranand @jredfield 

This makes no sense to me; why are we not able to have admins be able to edit this? We have done this same functionality with a variety of other options. example the last billing cycle date.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Feb 10, 2020, 24:04

@kdjstudios @jredfield 

> This makes no sense to me; why are we not able to have admins be able to edit this?

We can give the right to edit it. But here the concern is different. We need some fixed/unique value to differentiate this fee from regular fees. Because at the time of transfer of the amount on the child account, we will look for the specific fee on which we will create a staged fee. That's why we need either a fixed code or the fixed label on the fee.

> We have done this same functionality with a variety of other options. example the last billing cycle date.

Yes, right but their behavior/purpose is different. This staged fee behavior is similar to Holiday additional fee.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 10, 2020, 10:16

@jredfield 

Are you understanding what is needing to occur here? I am a bit confused, so marking for discussion. Could you please review this with the team in the morning meeting and provide an update.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Feb 12, 2020, 08:31

@kdjstudios @jredfield 

I'll try to explain it in more depth

* For transferring the child balance to the parent account, we are using a static fee `7050`.
* Whenever there will be a request to transfer the child account balance, it will look for the additional fee with code `7050` to add the staged fee.
* If we allow the admin to change the code of this static fee, then it will become a regular additional fee.
* Now, when there will be a request to transfer the child account balance, it will look for the static `7050` fee, which now does not exist as the admin has already changed the original code. And we will have no way to figure out which additional fee to add as the staged fee for the transfer.

I hope this helps

**NOTE:** We can allow admin to change the label of this static additional fee.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 12, 2020, 13:17

@ranand @amishra @jredfield 

Thank you for the clarification. I now understand that how you are going to create the code to find the fee needs to look for an exact match with either the Code or the Name. Lets have the code not be editable by anyone then, but the name can be editable by only admins. Will that work?

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Feb 12, 2020, 23:34

@kdjstudios @jredfield 

Thank you for the confirmation.

We have completed the development & set up a branch on the merge2 server for your testing.

Please test it.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 17, 2020, 12:44

@amishra 

I think this may have been replaced with a different branch on this server, I just attempted to test this and it is not working at all.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Feb 18, 2020, 06:46

@kdjstudios @jredfield 

Thank you for the update.

We have  setup the right branch on the merge2 server.

Please test it.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 18, 2020, 09:42

This appears to be working for me. Marking for announcement.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Mar 2, 2020, 02:51

@kdjstudios @jredfield 

We have successfully released on both of the production servers.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 2, 2020, 12:55

Hello Team,

I just received the following message when trying to create a new account. Why are we displaying this message for new accounts? Where is the date coming from? I think this is the pop up message that should only display when we take an existing parent account and make it a child account, but idk where the date is coming from. 

![image](/uploads/d8f598cf8e381c112d443374851024bf/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Mar 2, 2020, 21:59

@kdjstudios @jredfield 

Thank you for the update.

We looked into this but did not show the pop-up when we set up a new account for the **Demo Site** under the **Demo Company Customer**. 

Here is the short video [URL](https://www.loom.com/share/679c7d6b74264740a9c88f1f0ab3e2ac)  

Please have a look and let us know what step we missed.

It would be great if you can provide the step in which you were performed.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 3, 2020, 08:16

@amishra @jredfield 

IDK what was in that video, but it did not appear to be anyone creating a new account. 

Please perform the following steps:

1. Log into SA Billing Production Internal
2. Select Demo site from site drop down.
3. Select Kyle Demo from customer drop down.
4. Click "Go"
5. Click "New Account"
6. Fill in Account Number, Account Name, Resource ID, and Site.
7. Click "Save", then Click "No Deposit".
The notification will present to you.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Mar 3, 2020, 08:28

@kdjstudios @jredfield 

Thank you for the additional information.

I was missed the last step 7. It was my fault that I did not click the submit button.


Although, we have found the issue. I will fix it tomorrow morning on the production server directly.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 3, 2020, 08:55

@amishra 

Thank you. 

EDIT: I am also getting this pop up ever time I go in and make any change and save on an already existing child account.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Mar 4, 2020, 24:35

@kdjstudios @jredfield 

Thank you for the update.

We have fixed the issue now the pop-up will show only once when we make a child account.  

Please have a look.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 4, 2020, 10:57

This now appears to be working correctly.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 4, 2020, 10:57

closed

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 6, 2020, 12:39

Juan wrote:

> I am trying to update the billing for one our clients but getting prompt (attached). Please advise, thank you.
> 
> Account is not being moved or being set up for parent/child.

![ss3.6.20](/uploads/f6f1d2bbc302be6e2d65c9d0e5c2b076/ss3.6.20.PNG)

I wrote:

> May you please provide the account name and number this is occurring on?

Juan wrote:

> Carlos R. Hamilton III MD PA

@amishra @smasih @ranand @rranjan @jredfield 

This is occurring again!!! What is going on? Why is this popup showing when it should not be!?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 6, 2020, 12:39

reopened

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Mar 8, 2020, 22:52

@kdjstudios @jredfield 

Thank you for the update.

We looked into this but unable to replicate the issue.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Mar 9, 2020, 07:39

@kdjstudios @jredfield 

Can you please provide us some additional information or the steps the user has taken, so that we can replicate the issue? 

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 9, 2020, 08:41

I wrote:

> The support team was not able to reproduce this issue. May you please try editing the account again and let us know if you have any issues?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 11, 2020, 13:32

Juan wrote:

> Sorry was out of the office, still getting the same message. Screen cap attached, please advise if it's ok for me to save or if I should hold off.
[SA+Hosted+and+Managed+Services+-+Mozilla+Firefox+3_11_2020+10_06_47+AM.zip](/uploads/e812b06c149e703f1369190ddfe795d5/SA+Hosted+and+Managed+Services+-+Mozilla+Firefox+3_11_2020+10_06_47+AM.zip)

@ranand @amishra 

Please look into why this is still occurring.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Mar 12, 2020, 24:53

@kdjstudios @jredfield 

Thank you for the additional information.

Now, we found the issue and directly fixed on the production server.

Issue was, when a user does not access the parent account attrubute.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 12, 2020, 08:12

@amishra 

Thank you. Closing ticket.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 12, 2020, 08:12

closed
